lib/Mojo/WebSocketProxy/Backend/ConsumerGroups.pm

```diff
@@ -212,15 +212,26 @@ sub call_rpc {
     my $after_got_rpc_response_hooks  = delete($req_storage->{after_got_rpc_response})  || [];
     my $before_call_hooks             = delete($req_storage->{before_call})             || [];
     my $rpc_failure_cb                = delete($req_storage->{rpc_failure_cb});
+    my $rpc_timeout_extend_offset     = delete($req_storage->{rpc_timeout_extend_offset});
+    my $rpc_timeout_extend_percentage = delete($req_storage->{rpc_timeout_extend_percentage});
+
     # stream category which message should be assigned to
     $req_storage->{category} = $self->queue_separation_enabled && $req_storage->{category} ? $req_storage->{category} : DEFAULT_CATEGORY_NAME;
 
     foreach my $hook ($before_call_hooks->@*) { $hook->($c, $req_storage) }
 
+    my $category_deadline = $self->_rpc_category_timeout($req_storage->{category});
     my $category_timeout = $self->_rpc_category_timeout($req_storage->{category});
 
+    if ($rpc_timeout_extend_percentage) {
+        $category_timeout += $category_timeout * ($rpc_timeout_extend_percentage / 100);
+    }
+    if ($rpc_timeout_extend_offset) {
+        $category_timeout += $rpc_timeout_extend_offset;
+    }
+
     my $block_response = delete($req_storage->{block_response});
-    my ($msg_type, $request_data) = $self->_prepare_request_data($c, $req_storage, $category_timeout);
+    my ($msg_type, $request_data) = $self->_prepare_request_data($c, $req_storage, $category_deadline);
```

Comment:
why use deadline in this line and timeout in the next line?

Reply:
Because the timeout and the deadline may now be different values i.e bws can be forced to wait longer than the deadline for the response. So if a rpc worker picks up a job with 1us remaining and takes 500ms to respond then bws will already have responded a timeout error to the client which means the work was wasted and will be submitted again, most likely at a time we are under heavy load so we can allow an offset of 500mS between redis deadline and how long bws will wait for the reply.

Reply:
Thanks for your explanation